src/infra/flow-system-inmem/src/services/flow/flow_service_inmem.rs

```diff
@@ -192,39 +195,128 @@ impl FlowServiceInMemory {
             .try_get_flow_schedule(flow_key);
 
         if let Some(active_schedule) = maybe_active_schedule {
-            self.enqueue_auto_polling_flow(start_time, flow_key, &active_schedule)
+            self.enqueue_scheduled_auto_polling_flow(start_time, flow_key, &active_schedule)
                 .await?;
         }
 
         Ok(())
     }
 
     #[tracing::instrument(level = "trace", skip_all, fields(?flow_key, ?schedule))]
-    async fn enqueue_auto_polling_flow(
+    async fn enqueue_scheduled_auto_polling_flow(
         &self,
         start_time: DateTime<Utc>,
         flow_key: &FlowKey,
         schedule: &Schedule,
     ) -> Result<FlowState, InternalError> {
-        let trigger = FlowTrigger::AutoPolling(FlowTriggerAutoPolling {});
+        self.trigger_flow_common(
+            start_time,
+            flow_key,
+            FlowTrigger::AutoPolling(FlowTriggerAutoPolling {}),
+            |flow_run_stats: &FlowRunStats| {
+                schedule.next_activation_time(start_time, flow_run_stats.last_success_time)
+            },
+        )
+        .await
+    }
+
+    #[tracing::instrument(level = "trace", skip_all, fields(?flow_key))]
+    async fn enqueue_auto_polling_flow_unconditionally(
+        &self,
+        start_time: DateTime<Utc>,
+        flow_key: &FlowKey,
+    ) -> Result<FlowState, InternalError> {
+        // Very similar to manual trigger, but automatic reasons
+        self.trigger_flow_common(
+            start_time,
+            flow_key,
+            FlowTrigger::AutoPolling(FlowTriggerAutoPolling {}),
+            |_: &FlowRunStats| start_time,
+        )
+        .await
+    }
 
+    async fn trigger_flow_common(
+        &self,
+        trigger_time: DateTime<Utc>,
+        flow_key: &FlowKey,
+        trigger: FlowTrigger,
+        new_activation_time_fn: impl FnOnce(&FlowRunStats) -> DateTime<Utc>,
+    ) -> Result<FlowState, InternalError> {
+        // Query previous runs stats to determine activation time
+        let flow_run_stats = self.flow_run_stats(flow_key).await?;
```

Comment:
Would it make sense to check `run_stats` only if the `trigger_time - now < throttling_period` ? Would that help to eliminate a repository lookup in a good percentage of trigger cases?